dkg/pedersen/mod.go

```diff
@@ -22,66 +22,85 @@ var suite = suites.MustFind("Ed25519")
 type Pedersen struct {
 	privKey kyber.Scalar
 	mino    mino.Mino
-	rpc     mino.RPC
 	factory serde.Factory
 }
 
 // NewPedersen returns a new DKG Pedersen factory
-func NewPedersen(privKey kyber.Scalar, m mino.Mino) (*Pedersen, error) {
-	m, err := m.MakeNamespace("dkg")
-	if err != nil {
-		return nil, xerrors.Errorf("failed to create namespace: %v", err)
+func NewPedersen(privKey kyber.Scalar, m mino.Mino) *Pedersen {
+	factory := NewMessageFactory(suite, m.GetAddressFactory())
+
+	return &Pedersen{
+		privKey: privKey,
+		mino:    m,
+		factory: factory,
 	}
+}
 
-	factory := NewMessageFactory(suite, m.GetAddressFactory())
+// Listen implements dkg.DKG. It must be called on each node that participates
+// in the DKG. Creates the RPC.
+func (s *Pedersen) Listen() (dkg.Actor, error) {
 
-	h := NewHandler(privKey, m.GetAddress(), factory)
+	h := NewHandler(s.privKey, s.mino.GetAddress(), s.factory)
 
-	// TODO: should be done only in the Listen func.
-	rpc, err := m.MakeRPC("pedersen", h)
+	rpc, err := s.mino.MakeRPC("dkg", h)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to create RPC: %v", err)
 	}
 
-	return &Pedersen{
-		privKey: privKey,
-		mino:    m,
-		rpc:     rpc,
-		factory: factory,
-	}, nil
+	a := &Actor{
+		rpc:      rpc,
+		factory:  s.factory,
+		startRes: h.startRes,
+	}
+
+	return a, nil
 }
 
-// Listen implements dkg.DKG. It allows one to start a new DKG protocol.
-func (s *Pedersen) Listen(players mino.Players, pubKeys []kyber.Point, t int) (dkg.Actor, error) {
+// Actor allows one to perform DKG operations like encrypt/decrypt a message
+//
+// - implements dkg.Actor
+type Actor struct {
+	rpc      mino.RPC
+	factory  serde.Factory
+	startRes *startResult
+}
+
+// Setup implement dkg.Actor. It initializes the DKG.
+func (a *Actor) Setup(players mino.Players, pubKeys []kyber.Point, threshold int) error {
+
 	if players.Len() != len(pubKeys) {
-		return nil, xerrors.Errorf("there should be as many players as "+
-			"pubKey: %d := %d", players.Len, len(pubKeys))
+		return xerrors.Errorf("there should be as many players as "+
+			"pubKey: %d := %d", players.Len(), len(pubKeys))
+	}
+
+	if a.startRes.distrKey != nil {
```

Comment:
what if startRes is null ?

Reply:
I would expect the _startRes_ pointer to be nil until a setup is done. It's weird that it is set but with missing data.

Reply:
That's what I did first, but since the actor references it, if its nil when it references it, it won't be updated https://play.golang.org/p/1b-ignxqpGP
Thus, this is why I must initialize it first.

Reply:
Oh I see, I miss that. That's actually wrong because you only lock on the handler side but not on the actor one. That's why I was proposing a independent structure because it's much easier to manage and test. That's kinda what you did with the startRes thing but it's half-way done. In that case, it's having something like the following:
```go
type state struct {...}

func (s state) Done() bool { return s.isDone }

func (s state) GetDistKey() kyber.Point { return s.key }
```

Reply:
with the locking and stuff